File: code/287/solution.cpp
#include "../template/bigint.cpp"
#include "../template/china_reminder.cpp"
#include "../template/euler.cpp"
#include "../template/prime.cpp"
#include "../template/prime_cnt.cpp"
#include <iostream>/*{{{*/
#include <map>
#include <math.h>
#include <set>
#include <vector>
const long long e2 = 100;
const long long e3 = e2 * 10;
const long long e4 = e3 * 10;
const long long e5 = e4 * 10;
const long long e6 = e4 * e2;
const long long e7 = e6 * 10;
const long long e8 = e7 * 10;
long long mod      = 1e9 + 7;
using namespace std; /*}}}*/

/*add{{{*/
long long base_add(long long a, long long b) {
    a %= mod;
    b %= mod;
    return (a + b) % mod;
}
long long add(long long a, long long b) { return base_add(a, b); }
long long add(initializer_list<long long> args) {
    long long ret = 0;
    for (auto i : args)
        ret = base_add(ret, i);
    return ret;
}
/*}}}*/

/*{{{mul*/
long long base_mul(long long a, long long b) {
    a %= mod;
    b %= mod;
    long long ret = 0;
    while (b) {
        if (b & 1)
            ret = (ret + a) % mod;
        b >>= 1;
        a = (a + a) % mod;
    }
    return ret;
}
long long mul(long long a, long long b) { return base_mul(a, b); }
long long mul(initializer_list<long long> args) {
    long long ret = 1;
    for (auto i : args)
        ret = base_mul(ret, i);
    return ret;
} /*}}}*/

/*{{{qpow*/
long long qpow(long long a, long long b) {
    long long ret = 1;
    while (b) {
        if (b & 1)
            ret = mul({ret, a});
        a = mul({a, a});
        b >>= 1;
    }
    return ret;
} /*}}}*/
const int N           = 24;
const long long limit = 1LL << (2 * N - 2);
bool check(long long x, long long y) {
    // return true if black; otherwise false
    long long xx = x - (1 << (N - 1));
    long long yy = y - (1 << (N - 1));
    return xx * xx + yy * yy <= limit;
}
int dfs(long long x, long long y, long long length) {
    if (length == 1)
        return 2;
    long long xx = x + length - 1, yy = y + length - 1, sub_len = length >> 1;
    int cnt = 0;
    if (check(x, y))
        cnt++;
    if (check(xx, y))
        cnt++;
    if (check(x, yy))
        cnt++;
    if (check(xx, yy))
        cnt++;
    if (check(x + sub_len - 1, y + sub_len - 1))
        cnt++;

    if (cnt == 5 || cnt == 0)
        return 2;
    return 1 + dfs(x, y + sub_len, sub_len) + dfs(x + sub_len, y + sub_len, sub_len) + dfs(x, y, sub_len) + dfs(x + sub_len, y, sub_len);
}
int main() {
    cout << dfs(0, 0, 1 << N) << endl;
    return 0;
}

File: code/287/solution.go
package main /*{{{*/

import (
	"fmt"
	"reflect"
)

const (
	e4  = int64(10000)
	e5  = int64(100000)
	e6  = int64(1000000)
	e7  = int64(10000000)
	e8  = e4 * e4
	mod = int64(1e9 + 7)
)

func add(list ...interface{}) int64 { /*{{{*/
	if len(list) == 0 {
		return 0
	}
	ret := int64(0)
	for _, i := range list {
		buf := int64(0)
		switch reflect.TypeOf(i).String() {
		case "int":
			buf = int64(i.(int))
		case "int64":
			buf = i.(int64)
		}
		ret = (ret + buf) % mod
	}
	return ret
} /*}}}*/
func mul(list ...interface{}) int64 { /*{{{*/
	if len(list) == 0 {
		return 0
	}
	ret := int64(1)
	for _, i := range list {
		buf := int64(0)
		switch reflect.TypeOf(i).String() {
		case "int":
			buf = int64(i.(int))
		case "int64":
			buf = i.(int64)
		}
		buf %= mod
		base := ret
		ret = 0
		for buf > 0 {
			if (buf & 1) == 1 {
				ret = add(ret, base)
			}
			base = add(base, base)
			buf >>= 1
		}
	}
	return ret
} /*}}}*/
func qpow(a, b int64) int64 { /*{{{*/
	ret := int64(1)
	for b > 0 {
		if (b & 1) == 1 {
			ret = mul(ret, a)
		}
		a = mul(a, a)
		b >>= 1
	}
	return ret
} /*}}}*/
/*}}}*/

const N = int64(24)
const limit = int64(1 << (2*N - 2))

func check(x, y int64) int64 {
	xx := x - (1 << (N - 1))
	yy := y - (1 << (N - 1))
	if xx*xx+yy*yy <= limit {
		return 1
	}
	return 0
}
func dfs(x, y, length int64) int64 {
	if length == 1 {
		return 1
		return 2
	}
	xx, yy, sub_len := x+length-1, y+length-1, length>>1
	cnt := check(x, y) + check(xx, y) + check(x, yy) + check(xx, yy) + check(x+sub_len-1, y+sub_len-1)
	if cnt == 5 || cnt == 0 {
		return 1
		return 2
	}
	return 1 + dfs(x, y+sub_len, sub_len) + dfs(x+sub_len, y+sub_len, sub_len) + dfs(x, y, sub_len) + dfs(x+sub_len, y, sub_len)
}
func main() {
	fmt.Println(dfs(0, 0, 1<<N))
}
